Fix construction of Effect.Unhandled in bytecode

dra27 opened the pull request:
`caml_make_unhandled_effect` can allocate, so the call needs to ensure that the virtual machine is set-up for the GC.

The original segfault reported in #11669 arises from the `PERFORM` path - I've assumed that the `REPERFORMTERM` path could suffer in the same way. I couldn't re-trigger @shym's original failure with this patch.

gasche commented on runtime/interp.c:
```diff
@@ -1326,7 +1326,9 @@ do_resume: {
       if (parent_stack == NULL) {
+        Setup_for_gc;
         accu = caml_make_unhandled_effect_exn(accu);
+        Restore_after_gc;
```
This should be `Setup_for_c_call`. `Setup_for_gc` is incorrect here as it will save and restore the `accu` parameter, un-doing the work of line 1330.

(Is this true? If it happens as I described above it should have been caught by testing.)

dra27 replied:
Hmm, not sure how that's got through - the first version was `Setup_for_c_call` and I then convinced myself that was the wrong the pairing. I'll have a look at this again tomorrow.

dra27 replied:
Fixed to use `Setup_for_c_call` - the difference with the `pc` storage here doesn't matter

gasche commented on runtime/interp.c:
```diff
@@ -1372,7 +1374,9 @@ do_resume: {
         resume_fn = raise_unhandled_effect;
+        Setup_for_gc;
         resume_arg = caml_make_unhandled_effect_exn(eff);
+        Restore_after_gc;
```
I find `Setup_for_gc` around a function that is not an immediate GC call rather weird / unidiomatic, but I think that it is correct -- it is needed to preserve the `accu` global that is set on line 1375. Another option would be to use `Setup_for_c_call` instead, but move the `accu = ` on line 1375 after the allocating call, so that `accu` needs not be preserved.

dra27 replied:
The other thing I considered - given that `caml_make_unhandled_effect_exn` is a refactoring which is only present to allow the bytecode interpreter to do this was either to convert it to a `Make_unhandled_effect_exn` macro (and so use `Alloc_small` with `Enter_gc` directly) or a `Caml_inline` function in `interp.c`.

gasche replied:
I thought about this as well, but it is also used by `caml_raise_unhandled_effect` which is itself used by runtime files through `caml_c_call`.

dra27 replied:
What I meant was to merge `caml_raise_unhandled_effect` and `caml_make_unhandled_effect_exn` back together (for native code) and then hand-roll a bytecode version in interp.c - the allocation function differs in each case. It wasn't clear to me that `Setup_for_c_call` is necessarily the _right_ thing here.

dra27 replied:
`Setup_for_gc` .. `Restore_after_gc` is used in `Enter_gc` - I think is a bracketting it's more "setup for the _possibility_ of the gc being called", i.e. it brackets something which allocates?

gasche replied:
Yes, but `Enter_gc` itself is only used in `Alloc_small` invocations and nowhere else. You are breaking this pattern by using it to call an apparently-arbitrary C function (this function is not special, many C functions can call the OCaml GC, including all the unknown functions in the `C_CALL<n>` instructions).

> What I meant was to merge `caml_raise_unhandled_effect` and `caml_make_unhandled_effect_exn` back together (for native code) and then hand-roll a bytecode version in interp.c - the allocation function differs in each case.

Yes, I think that would work. (Note: I'm slightly annoyed by `caml_raise(...)` taking a `value` temporary without rooting; if you rewrite this function I hope this will go away.)

dra27 replied:
Hmm, I just had a look at rewriting the function, but it's really not as neat as I initially thought it might be - there's the exception caching, which either has to be exposed (slightly hairy, given the atomics, etc.) or _duplicated_.

This function is not special, but the function _call_ is special - as far as I can see, it's the only instance in the entire interpreter loop where we call a function which may allocate **and** where `accu` needs to be preserved across it. It looks to me (possibly incorrectly), that _all_ the other instances are in fact `Setup_for_gc` unless they can either be proven not to be needed at all (several comments at the start of the interpreter around use of `caml_alloc_shr`) or can be _optimised_ to `Setup_for_c_call` where we know that either `accu` is registered as a root by the function being called and so does not need to be preserved (C_CALL) cases or is not live/already registered (exception raising cases).

xavierleroy replied:
There may be a simpler way to go about all this, let me think about it.  Meanwhile, I'm puzzled by the line
```
accu = caml_continuation_use(cont);
```
~~It looks like the value of `accu` is not used by the code at label `do_resume`.  Do you agree?  What does this mean? (to ignore the return value of `caml_continuation_use`)?~~ I misread the code, `accu` is used and it all makes sense.

dra27 committed:
Fix construction of Effect.Unhandled in bytecode

caml_make_unhandled_effect can allocate, so the call needs to ensure
that the virtual machine is set-up for the GC.

Co-authored-by: Xavier Leroy <xavier.leroy@inria.fr>

xavierleroy commented on runtime/interp.c:
```diff
@@ -1372,7 +1376,9 @@ do_resume: {
       if (parent == NULL) {
         accu = caml_continuation_use(cont);
         resume_fn = raise_unhandled_effect;
+        Setup_for_gc;
         resume_arg = caml_make_unhandled_effect_exn(eff);
+        Restore_after_gc;
```
I think there's another problem here: `resume_fn` is not a root, so if a GC occurs just after, the closure `raise_unhandled_effect` can be moved and `resume_fn` is invalidated.

Here is a suggestion that addresses both this problem and the problem with `accu`:

```suggestion
        Setup_for_c_call;
        resume_arg = caml_make_unhandled_effect_exn(eff);
        accu = caml_continuation_use(cont);
        Restore_after_c_call;
        resume_fn = raise_unhandled_effect;

```
As far as I can see, `caml_continuation_use` doesn't allocate in the heap, at least when it returns normally.  However, it can raise an exception, which is why I moved it inside the Setup/Restore section.  (For an esception raised by C to be trapped by the bytecode interpreter, the `domain_state->current_stack->sp` must contain the correct `sp` value; the `Setup_for_*` macros take care of that.)

dra27 replied:
That gets those proverbial two birds with one stone, thanks! 🙂

It's clearly better to have the code such that we know `resume_fn` can't be invalidated. Just for my understanding, I think it's the case (without a compactor) that it couldn't have moved before, though - `raise_unhandled_effect` is a global root and the value is set-up when the interpreter is initialised and before the first minor collection at startup. It would therefore have always been in a fixed place in the major heap when the interpreter actually started the interpreting the program?

xavierleroy replied:
You're probably right, because there's no compactor and a minor GC moved the closure to the major heap earlier.  But this is a complicated reasoning that I don't want to make, ever.  Any heap-allocated value can move during any allocation, that's a simple rule.

xavierleroy reviewed (commented):
Some more juggling of operations to be safe w.r.t. GC and exceptions...

gasche commented on runtime/interp.c:
```diff
@@ -1326,7 +1326,11 @@ do_resume: {
       check_trap_barrier_for_effect (domain_state);
       if (parent_stack == NULL) {
+        /* Setup_for_c_call, so accu isn't clobbered by Restore_after_gc.
+           caml_make_unhandled_effect handles registering accu. */
```
I'm not sure this comment is necessary (because in my mind it wouldn't be natural to use `Setup_for_gc` here so justifications for not doing it are not necessary; this code is similar to the `C_CALL<n>` cases which need no justification), but having more comments than necessary is better than the other way around.

dra27 replied:
I agree - I don't think this adds much in the final version of the code. I needed to rebase to update Changes, so I've deleted it.

gasche reviewed (approved):
I believe that the fix is now correct and the code looks idiomatic to me. Approved.

I wonder if there are other issues lurking with the new effect-handler code in interp.c, that hasn't been written by interp.c experts. @dra27 or @xavierleroy, has one of you per chance considered looking at the rest for similar issues?

dra27 commented:
I know @kayceesrk and @stedolan place a lot of trust in the effects tests in the testsuite for really torturing the implementations; @jmid and @shym (et al!)'s testing is very fruitfully leading us to instances of unsafe use of unsafe functions and, as here, missed interactions with the GC on top of that. I've certainly stared at https://github.com/ocaml/ocaml/commit/c77ce613de51671f425d6a3e30ac5fd05417f286 quite a lot as part of this, but I'm not sure that we necessarily should worry (too much) in the absence of further bug reports for picking over it further.

xavierleroy commented:
Appveyor fails the testsuite with
```
** Fatal error: Cannot run cygpath -m "libgcc" "libgcc.lib" [very long list of arguments]
```
but I'm going to assume this is unrelated to this PR.

xavierleroy commented:
Cherry-picked  to 5.0: d225a54141

kayceesrk commented:
Thanks for the fix. The code looks correct to me. It was an oversight not to carefully consider the interaction with the GC in the original PR. 

I do wonder whether there are ways to test the GC interactions using (possibly custom) sanitisers of some kind.
